### ISSUE TITLE
Last hurdle

### ISSUE THREAD

**Issue by Bartekus** [Araphel/meatier-old#1]
Ok @mattkrick at this point I've smoothed out all the problems with azk.io and we are almost there, however I'm currently running into this issue:

```
root@f071e1d2fc30:/azk/r3stack# npm run build:db
npm info it worked if it ends with ok
npm info using npm@3.3.12
npm info using node@v5.2.0
npm info lifecycle r3stack@1.0.0-alpha~prebuild:db: r3stack@1.0.0-alpha
npm info lifecycle r3stack@1.0.0-alpha~build:db: r3stack@1.0.0-alpha

> r3stack@1.0.0-alpha build:db /azk/r3stack
> node ./src/server/database/setupDB.babel.js

Creating a pool connected to rethinkdb.dev.azk.io:32776
Draining the pool connected to rethinkdb.dev.azk.io:32776
npm info lifecycle r3stack@1.0.0-alpha~postbuild:db: r3stack@1.0.0-alpha
npm info ok
root@f071e1d2fc30:/azk/r3stack# npm run build:client
npm info it worked if it ends with ok
npm info using npm@3.3.12
npm info using node@v5.2.0
npm info lifecycle r3stack@1.0.0-alpha~prebuild:client: r3stack@1.0.0-alpha
npm info lifecycle r3stack@1.0.0-alpha~build:client: r3stack@1.0.0-alpha

> r3stack@1.0.0-alpha build:client /azk/r3stack
> NODE_ENV=production webpack --colors --config ./webpack/prod.babel.js

Hash: a19f19da1b055f4aab9a
Version: webpack 2.0.2-beta
Time: 36649ms
                               Asset     Size  Chunks             Chunk Names
37681ed619fcde1be52da659ee7288ff.jpg   243 kB          [emitted]
           0_bc60c6ad279a86d47deb.js  27.6 kB    0, 4  [emitted]
           1_be618267abb21d5f8b9f.js  36.9 kB    1, 4  [emitted]
           2_9077508edbcc1c86d2c1.js   138 kB    2, 4  [emitted]
           3_52028c45d862583d7bbb.js   459 kB    3, 4  [emitted]
    manifest_dbf84eff6deaa906b2ab.js  1.14 kB       4  [emitted]  manifest
         app_f5d1b117bacfd3a66d4f.js   499 kB    5, 4  [emitted]  app
      vendor_28a30fcb97bda3f33e1b.js   504 kB    6, 4  [emitted]  vendor
   [2] multi app 40 bytes {5} [built]
   [3] multi vendor 52 bytes {6} [built]
    + 1119 hidden modules
npm info lifecycle r3stack@1.0.0-alpha~postbuild:client: r3stack@1.0.0-alpha
npm info ok
root@f071e1d2fc30:/azk/r3stack# npm run build:server
npm info it worked if it ends with ok
npm info using npm@3.3.12
npm info using node@v5.2.0
npm info lifecycle r3stack@1.0.0-alpha~prebuild:server: r3stack@1.0.0-alpha
npm info lifecycle r3stack@1.0.0-alpha~build:server: r3stack@1.0.0-alpha

> r3stack@1.0.0-alpha build:server /azk/r3stack
> NODE_ENV=production webpack --colors --config ./webpack/server.babel.js

Hash: b82c433bddb8c239db63
Version: webpack 2.0.2-beta
Time: 24081ms
                               Asset     Size  Chunks             Chunk Names
37681ed619fcde1be52da659ee7288ff.jpg   243 kB          [emitted]
                        prerender.js   987 kB       0  [emitted]  prerender
                       prerender.css  31.1 kB       0  [emitted]  prerender
    + 779 hidden modules
Child extract-text-webpack-plugin:
        + 2 hidden modules
Child extract-text-webpack-plugin:
        + 2 hidden modules
Child extract-text-webpack-plugin:
        + 2 hidden modules
Child extract-text-webpack-plugin:
        + 2 hidden modules
Child extract-text-webpack-plugin:
        + 2 hidden modules
Child extract-text-webpack-plugin:
        + 2 hidden modules
Child extract-text-webpack-plugin:
        + 2 hidden modules
Child extract-text-webpack-plugin:
        + 2 hidden modules
Child extract-text-webpack-plugin:
        + 2 hidden modules
Child extract-text-webpack-plugin:
        + 2 hidden modules
Child extract-text-webpack-plugin:
        + 3 hidden modules
Child extract-text-webpack-plugin:
        + 5 hidden modules
Child extract-text-webpack-plugin:
        + 3 hidden modules
Child extract-text-webpack-plugin:
        + 3 hidden modules
Child extract-text-webpack-plugin:
        + 3 hidden modules
Child extract-text-webpack-plugin:
        + 3 hidden modules
npm info lifecycle r3stack@1.0.0-alpha~postbuild:server: r3stack@1.0.0-alpha
npm info ok
root@f071e1d2fc30:/azk/r3stack# npm run prod
npm info it worked if it ends with ok
npm info using npm@3.3.12
npm info using node@v5.2.0
npm info lifecycle r3stack@1.0.0-alpha~preprod: r3stack@1.0.0-alpha
npm info lifecycle r3stack@1.0.0-alpha~prod: r3stack@1.0.0-alpha

> r3stack@1.0.0-alpha prod /azk/r3stack
> NODE_ENV=production node ./src/server/server.babel.js

   [Busy] Launching SocketCluster
   >> Broker PID: 173
1453781051150 - Origin: Broker (PID 173)
   [Error] Error: listen EPERM /azk/r3stack/socketcluster/Meatier_4fef3560d7/b0
    at Object.exports._errnoException (util.js:856:11)
    at exports._exceptionWithHostPort (util.js:879:20)
    at Server._listen2 (net.js:1225:19)
    at listen (net.js:1274:10)
    at Server.listen (net.js:1364:5)
    at EventEmitter.self.listen (/azk/r3stack/node_modules/ncom/index.js:139:19)
    at Object.<anonymous> (/azk/r3stack/node_modules/sc-broker/server.js:505:10)
    at Module._compile (module.js:399:26)
    at Object.Module._extensions..js (module.js:406:10)
    at Module.load (module.js:345:32)
```

Any ideas of top of your head what might be causing this?

I tried running the command with sudo as per:
>https://github.com/SocketCluster/socketcluster/issues/117

But that didn't make a difference.
Now I'm looking into this:
>https://github.com/SocketCluster/socketcluster/issues/142

But if you have any ideas, just shoot

**Comment by Bartekus** [comment]
I tried running the command with sudo as per:
> https://github.com/SocketCluster/socketcluster/issues/117

But as per this:
>https://github.com/SocketCluster/socketcluster/issues/142

I'm converting it to use ES5 as it could be exactly that.

**Comment by mattkrick** [comment]
you need to recursively chmod that entire folder. in my experience, running sudo can fuck shit up since if you create something as root you'll lock yourself out of it (can't tell you how many times i did that with mongo until i learned my lesson)

**Comment by Bartekus** [comment]
My sentiment exactly, one does not simply run sudo as they say ;)
So yeah, I recreated folder structure by just erasing everything minus the setting and starting from scratch. I'm trying to at all time simulate what a normal user can experience when dealing with r3stack.

**Comment by Bartekus** [comment]
Here are the facts: currently I run my rethinkdb using docker and orchestrated by azk.io using command:

>azk start rethinkdb 

All works and I get the azk to bring the db up and confirm its settings:
```
┌───┬───────────┬───────────┬─────────────────────────────┬────────────────────────────┬────────────────┐
│   │ System    │ Instances │ Hostname/url                │ Instances-Ports            │ Provisioned    │
├───┼───────────┼───────────┼─────────────────────────────┼────────────────────────────┼────────────────┤
│ ↑ │ rethinkdb │ 1         │ http://rethinkdb.dev.azk.io │ 1-http:32773, 1-data:32772 │ -              │
│   │           │           │                             │ 1-29015:32771              │                │
├───┼───────────┼───────────┼─────────────────────────────┼────────────────────────────┼────────────────┤
│ ↓ │ r3stack   │ 0         │ r3stack.dev.azk.io          │ -                          │ 34 minutes ago │
└───┴───────────┴───────────┴─────────────────────────────┴────────────────────────────┴────────────────┘
```

then I modify my rethink.config.js with the setting that it give me:
```
export default {
  host: 'rethinkdb.dev.azk.io',
  port: 32772,
  db: process.env.NODE_ENV === 'testing' ? 'ava' : 'meatier',
  min: process.env.NODE_ENV === 'production' ? 50 : 3,
  buffer: process.env.NODE_ENV === 'production' ? 50 : 3
}
```

and now I can just run it using the usual:

>npm run quickstart

everything work perfectly n/p. So I confirmed that bot the dockerized db is operational and accessible and also that from the osx environment meatier starts without a hitch too.

ok now we gonna purge the database, and erase build and socketcluster folder;
then with a clean slate, we enter the docker environment:

>azk shell r3stack

once inside, just like before we will go with the streamlined quickstart:

>npm run quickstart

Just like before, it do connect to db, sets structure, just the client, then the server, then the broker comes on and boom, it pukes out failing to connect:

```
[3]    [Busy] Launching SocketCluster
[3]    >> Broker PID: 578
[3] 1453982431582 - Origin: Broker (PID 578)
[3]    [Error] Error: listen EPERM /azk/r3stack/socketcluster/Meatier_4fef3560d7/b0
[3]     at Object.exports._errnoException (util.js:856:11)
[3]     at exports._exceptionWithHostPort (util.js:879:20)
[3]     at Server._listen2 (net.js:1225:19)
[3]     at listen (net.js:1274:10)
[3]     at Server.listen (net.js:1364:5)
[3]     at EventEmitter.self.listen (/azk/r3stack/node_modules/ncom/index.js:139:19)
[3]     at Object.<anonymous> (/azk/r3stack/node_modules/sc-broker/server.js:505:10)
[3]     at Module._compile (module.js:399:26)
[3]     at Object.Module._extensions..js (module.js:406:10)
[3]     at Module.load (module.js:345:32)
```

>EPERM

So this is the breakdown: 
We can install it from the osx, we cannot install it from the virtualbox/ unless with sudo, crime #1
If we install it from osx, we can run it in osx and we get all the functionality, it just works. but if we go into the shell and try to run it from there, it will end with the EPERM problem

if we install it from the shell with sudo, just like in the previous example it's still falls flat on it's tits in exact place, altho it at least installs. 

I found some clues here:
https://github.com/npm/npm/issues/3565

and given all that, plus the fact that currently azk.io does not support windows, I don't believe there is much point in attempting to use azk.io beyond the database setup, which it does splendidly well so that atlas we can make it very easy to run thinkdb which along with the rest of the deployment can make meatier quite easy to install and startup/use for osx and linux users.... windows not so much... 

orchestration can still be done thru dockerizng the build using docker and not azk.io once the build is ready to be deployed, and using azk to actually deploy it, so extra step in between that can be scripted. I think it pretty good solution, or at least with my limited knowledge it is. But I'd like to hear your opinion about it.

**Comment by mattkrick** [comment]
@jondubois Ever run into this when deploying socketCluster? Given the EPERM, I think it's locking itself out of the temp folder used by websocket.

**Comment by jondubois** [comment]
@mattkrick @Bartekus This issue can happen if you ran SC as root and then, using the same appName, you run it again as a user with lower privileges.

A new directory is created to hold Unix domain socket handles inside the tmp/ folder based on the appName you provide. So if you run SC as root, it will create a directory owned by root, then if you try to run the same app again but this time as a regular user, it will try to create domain socket handles inside the same directory (owned by root), so it would fail with this error.

Unix domain sockets are using by SC to communicate between processes.
Maybe the user name should be part of the directory name to avoid conflicts like this.

**Comment by mattkrick** [comment]
@jondubois thanks for the detailed response! PS, this repo is gonna make deploying a socketcluster website dead simple. you built the backend, I did the front end, bartekus is doing the devops. time to bring your tech to the masses :wink:

@Bartekus did you rimraf the tmp folder when you tried installing a second time? As an easy alternative, try changing the appName in the `server.js` to create a new dir. or chmod the entire thing. or just nuke the VM

**Comment by Bartekus** [comment]
I'm sorry for the delay gentlemen, I'm just taking a little bit of side time to bring araphel.org online, I should be done in few hours

**Comment by Bartekus** [comment]
ok, everything's reinstalled, starting from very clean system, there is nothing in the way of an error or some error artifact getting in the way. I grab the fresh copy of meatier, just finished setting azk control file and now I'm just confirming with the azk team what setting to go with for most optimal approach... I'll keep you guys posted on the progress

**Comment by Bartekus** [comment]
@jondubois @mattkrick 
ok gentlemen, got it now, and its 99.7% operational, so I'm just hammering out the last little inconsistency, mainly with graphql aspect but it shouldn't be long before I overcome this little bump.
 
Also, due to how GitHub operates and the goals that we want to achieve with the stack, a proper organization solely dedicated to the r3stack has been created, it uses the r3stack logo and has it's own url: r3stack.com 

So while owner invite has already been sent, I know that Matt is wholeheartedly on board, however since I never directly asked, I'm not sure how Jon feels about being one of the fathers of this undertaking, I certainly would love for him to participate, if only in membership; especially being that vast part of the project is due to his love and dedication to the open source community through his socketcluster and other contributions.

Additionally due to intend of full open nature and multi-national compliance, an i18n specification is especially desirable. As such I've have taken additional initiative and started to cooperate with [OpenCredo](http://opencredo.com) (Software Development Consultancy from London, UK) as they are currently developing such solution in form of [opencredo-react-boilerplate](https://github.com/opencredo/opencredo-react-boilerplate). 
So have a look at it Matt and let me know what you think. I've already been digesting the inner working of it while making some contributions and suggestion, but overall, I'm very impressed and see a lot of potential there. The standard and the quality behind the project is simply superb. They are in an essence putting the knowledge that they have gained over the last 6 or so projects that they have completed, into the essence of that boilerplate. The project itself has only few stars as it's very young and relatively unknown but once people discover the quality that it is build on and how it deals with the current problem, it will quickly overtake kriasoft/react-starter-kit.
So yeah I'm intending to spread the news about it, but not before I learn everything I can and integrate it to our r3stack.

Furthermore, while I do not believe it's possible at the moment, (for one, until they upgrade react-router to v2) I've made favorable contacts with [Structor](https://github.com/ipselon/structor) team (currently starting angel investor bid) and they are aware of our intend to use their technology and are very happy to assist us in that integration.

Ok, that's all for now, I'm going back to work but I'll keep you all posted.

PS. February 11 is the launch day, but I'm certain we will have everything ironed out in less than a week at this point seeing how complete the solution really is.

**Comment by Bartekus** [comment]
ok, it is done... I've forked the meatier, then applied the necessary changes, did the basic test and it all looks good gentlemen. 
Please try it out and let me know, I'll drop this branch as it's better to keep proper fork for ongoing development and such, gotta keep it proper and by the book as they say.

**Comment by jondubois** [comment]
This extremely cool what you guys are doing. I'm definitely in :)
I like the way this is all going with the Structor drag-and-drop UI builder too.
This could become a really powerful stack for application development.

It would be nice if we had a common chat channel on Gitter or Slack where we can chat and plan stuff. I have some ideas I would love to discuss.

**Comment by Bartekus** [comment]
ok, it's great to have you on board Jon, as per suggestion I've created two channels, one to deal directly with r3stack issues as they appear, while the one under Araphel, Derekh is going to be our main discussion room of the general direction in which we intend to take the project. As fathers of this undertaking, you have full authority over each and every aspect of the project. You are free to bring whoever you wish on board as well as make any suggestion, proposal or addition you feel would be of benefit to the overall target that we have. Which technically is also up to be established. I personally feel that frameworks tend to at first enable, but quickly descent to crippling the progress. As such, while we will provide whole solutions, our goal will not be to provide back-wards compatibility, or support the technology which has been replaced by better solutions. I believe our goal is to foster advancement thru construction of the best solutions, given the available tools and reflecting their use in solving of the problems that are current and most commonly faced by largest spectrum of community. That is not to say that we won't create off-shoots project that tackle specific problems in a specific manner. However we are not angular nor do we want to be ember. Our way is the way of the flow; we bend when wind blows, yet we stand firm when the sun shines. We adapt the mantra of the rails project in a way, to foster community of like minded individual. The 12Factor is our guidance, hacker ethics is our mantra, servitude to the community and improvement to quality of life is our objective. By hackers, for hackers, from one mind to the other, its the way of the araphel.
